Venstar air filter sensors

jhollowe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add two sensors to the venstar integration which present the filter replacement/usage schedule.
The thermostat keeps track of both how many days the filter has been installed and how many active
hours the filter has had air flow. Both values are reset when the filter life is reset on the thermostat after replacing the filter.

![chrome_V0OWqobsho](https://github.com/home-assistant/core/assets/2881268/194f1abb-6849-456c-ae2a-c30ae60088a7)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

frenck commented:
Oh hi there @jhollowe 👋 

This PR is now in draft for almost two months. Are you planning on moving this forward / mark ready for review?

../Frenck

jhollowe committed:
Add sensors for HVAC filter usage

The number of hours and the number of days are not linked.
The number of hours is a sum of the hours the filter has been in use (fan running). The days is just the number of days since the filter was reset.

joostlek commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -160,15 +178,15 @@ def native_value(self) -> int:
     def native_unit_of_measurement(self) -> str | None:
         """Return unit of measurement the value is expressed in."""
-        return self.entity_description.uom_fn(self.coordinator)
+        return self.entity_description.uom_fn(self.coordinator, self.sensor_name)
```
I think we can do this smarter, but let's await the other question first

joostlek replied:
To describe my issue with this btw, let's add a way to do static unit of measurements without adding all kinds of untyped parameters to a function.

Something like the default implementation would work

```py
if hasattr(self.entity_description, "native_unit_of_measurement"):
    return self.entity_description.native_unit_of_measurement
```
If you add this as the first line of the function, you could have 2 entity descriptions for the new entities and just set `native_unit_of_measurement`.

joostlek commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -58,6 +63,11 @@
     RUNTIME_OV: "Override",
 }
 
+CONSUMABLE_ATTRIBUTES = {
+    CONSUMABLE_FILTER_HOURS: {"name": "Filter Used Hours", "uom": UnitOfTime.HOURS},
+    CONSUMABLE_FILTER_DAYS: {"name": "Filter Used Days", "uom": UnitOfTime.DAYS},
+}
+
```
Now I'm reading the PR description again I'm assuming the days means "days this has been installed" and hours means "hours of active use". I think we should make this clearer in the names. Preferably the names should not contain the unit of measurement, because like said, that is changeable by the user and looks strange to suddenly measure "Filter Used Days" in weeks.

jhollowe replied:
I was mimicing how the existing "runtime" devices/attributes are.

joostlek replied:
How do you mean runtime?

jhollowe committed:
Update filter sensors' names

allout58 commented:
@jhollowe what is left to get this PR merged? I also use Venstar thermostats and would really appreciate being able to have Home Assistant notifications when I need to replace my air filters.

joostlek reviewed (changes requested):
My comments were not addressed yet nor replied to

jhollowe committed:
Split consumables entity descriptions and move names to translations

jhollowe committed:
Scale filterHours to match real-world time

It looks like the integer returned by the thermostat is actually 100 times the number of hours.

joostlek commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -99,11 +99,20 @@ async def async_setup_entry(
+                        coordinator, config_entry, RUNTIME_ENTITY, sensor_name
+                    )
+                )
+            entities.extend(
+                [
+                    VenstarSensor(coordinator, config_entry, description, sensor_name)
+                    for description in CONSUMABLE_ENTITIES
+                    if description.key == sensor_name
+                ]
+            )
```
```suggestion
            entities.extend(
                VenstarSensor(coordinator, config_entry, description, sensor_name)
                for description in CONSUMABLE_ENTITIES
                if description.key == sensor_name
            )
```

joostlek commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -224,6 +233,26 @@ def native_unit_of_measurement(self) -> str | None:
+        key="filterHours",
+        state_class=SensorStateClass.MEASUREMENT,
+        uom_fn=lambda _: UnitOfTime.HOURS,
+        value_fn=lambda coordinator, sensor_name: coordinator.runtimes[-1][sensor_name]
+        / 100,
```
Let's wrap the lambda in paranethesis

joostlek commented on homeassistant/components/venstar/strings.json:
```diff
@@ -25,6 +25,12 @@
   "entity": {
     "sensor": {
+      "filter_install_time": {
+        "name": "Filter Installation Time"
+      },
+      "filter_usage": {
+        "name": "Filter Usage"
```
Please use sentence case

jhollowe committed:
Address review comments/changes

joostlek commented:
Thanks!
